### PR TITLE
NEP: issue deprecation warning when creating ragged array (NEP 34)

### PR DESCRIPTION
**Pull request by mattip**
Redo gh-14794 ~but add a `np.allow_object` sentinel to retain legacy behaviour.~ 

Also see gh-15118 which is tangentially connected to this PR, where the `context` argument now serves dual purposes.

### CODE REVIEW

**Review by seberg — APPROVED** [review]
LGTM, just some nitpicks. Others may have opinions about using `context` for this purpose, but it is convenient. Mainly waiting for a clear decision the whole thing I guess.

**Comment by mattip** [comment]
waiting on gh-15083 to resolve the name and need for the singleton

**Comment by mattip** [comment]
It turns out downstream libraries (I checked pandas, scipy) adapted to NEP 34 even without the sentinel. Once this is ready and reviewed, I would like to try again to see if it causes less noise this time

File: numpy/core/tests/test_multiarray.py
Comment (seberg) [inline numpy/core/tests/test_multiarray.py]: `dtype` appears to be unused above? So its always object I think?

File: numpy/core/tests/test_deprecations.py
Comment (seberg) [inline numpy/core/tests/test_deprecations.py]:
```suggestion
    # 2019-11-29 1.19.0
```

File: numpy/core/_internal.py
Comment (seberg) [inline numpy/core/_internal.py]: Is this (still) necessary? I am not sure how it is related.

File: doc/release/upcoming_changes/template.rst
Comment (seberg) [inline doc/release/upcoming_changes/template.rst]: I think you changed that for the release notes? I am not sure if we still want it or not.

**Review by seberg — COMMENTED** [review]
LGTM (minor things). IIRC this is pretty much what we had before, right? We probably still need to (re-) add the release notes?

File: numpy/core/tests/test_multiarray.py
Comment (mattip) [reply numpy/core/tests/test_multiarray.py]: `dtype` is used in `_ragged_creation`, once as `object` and once as `None`. The `None` use triggers a `DeprecationWarning`

File: numpy/core/tests/test_deprecations.py
Comment (mattip) [reply numpy/core/tests/test_deprecations.py]: fixed

File: numpy/core/_internal.py
Comment (mattip) [reply numpy/core/_internal.py]: it is only tangentially related, in that `ctypes` is not very specific and so can be confused, especially since we have a `numerictypes` module. I would like to leave the change even though it is not needed by this PR, since it may prevent someone in the future spending too much time staring at an error. Although `_ctypes` is only marginally better. This is all a hack around a cpython issue.

File: doc/release/upcoming_changes/template.rst
Comment (mattip) [reply doc/release/upcoming_changes/template.rst]: yes, I now added a release note that needs the blank line

File: numpy/core/_internal.py
Comment (eric-wieser) [reply numpy/core/_internal.py]: Whoops, I hadnt considered numeri**ctypes""

File: numpy/core/src/multiarray/ctors.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Should this be 19 now?

File: numpy/core/src/multiarray/ctors.c
Comment (eric-wieser) [inline numpy/core/src/multiarray/ctors.c]: Should this include the word ragged?

File: numpy/core/tests/test_multiarray.py
Comment (seberg) [reply numpy/core/tests/test_multiarray.py]: Yes, but `_ragged_creation` always does both, so there is no need for it here.

File: numpy/core/src/multiarray/ctors.c
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: fixing

File: numpy/core/src/multiarray/ctors.c
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: fixing

File: numpy/core/tests/test_multiarray.py
Comment (mattip) [reply numpy/core/tests/test_multiarray.py]: fixing

File: numpy/core/src/multiarray/ctors.c
Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Looks like this was missed? I can fix it up though.

File: numpy/core/src/multiarray/ctors.c
Comment (seberg) [inline numpy/core/src/multiarray/ctors.c]: Now that Eric pointed out that we may want to include "ragged", I am thinking we should actually bloat this up a lot more. Maybe even create a `VisibleDeprecationWarning` (although that is a bit annoying).

The reason is that we are currently aiming the error at users who want to keep the old behaviour, when actually most users who see it are likely people who simply mistyped and thus have a bug. For them "ragged" is probably not helpful except as a google term.

It is a bit long I know, but am considering something like:
> You are attempting to create an array, or doing an array operation, with uneven shape (ragged). This is not supported and may be an input error such as `np.array([[1], [1, 2]])` where the first list is shorter than the second. If you wish to create an array containing sequences of different length you must create an array with `dtype=object` in the future.

**Review by seberg — APPROVED** [review]
Sorry for dragging this out, but I am considering if we should try to spell out/target the error message for the unexperienced user who likely just got it by accident. We could even add that other tools may exist to handle such data, but not sure about that?

File: numpy/core/src/multiarray/ctors.c
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: Changed to a `VisibleDeprecationWarning`, extended the message, rebased off master, and fixed the deprecation version in e11d57a.

Although in my mind the difference between a VisibleDeprecationWarning and an Error is negligible: emitting a VisibleDeprecationWarning is so annoying it means there is no real deprecation period, users will, in practice, have to fix their code immediately.

Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: Searching for ["ragged nested sequences numpy"](https://duckduckgo.com/?q=ragged+nested+sequences+numpy&t=canonical&ia=web) the first hit is NEP 34, so I think this text is good enough.

**Comment by mattip** [comment]
Should be squashed before merging so we can revert a single changeset.

File: numpy/core/src/multiarray/ctors.c
Comment (seberg) [inline numpy/core/src/multiarray/ctors.c]: `npy_cache_import` includes this if (and is guaranteed to be inlined)

File: numpy/core/src/multiarray/ctors.c
Comment (seberg) [inline numpy/core/src/multiarray/ctors.c]:
```suggestion
                    "ndarray from ragged nested sequences (which is a "
```

**Review by seberg — COMMENTED** [review]
Good point about the googlebility of "ragged nested sequences". I am OK with this, just thoughts. About the `VisibleDeprecationWarning` if anyone else has an opinion, please just note. I think I am still a little bit in favor: The assumption is that in the vast majority of (non-library) use cases, it currently is actually a silent error. If anyone thinks that is not true, we should first do the normal deprecation.

File: numpy/core/src/multiarray/ctors.c
Comment (seberg) [reply numpy/core/src/multiarray/ctors.c]: Ack, forgot to press submit: Maybe rather than the bracketed explanation, put in the example: `such as '[[1], [2, 3]]' where '[2, 3]' is longer than '[1]'`? I would hope that is more useful.

Not sure though, and I am fine with it as is (we can always revise as well)

File: numpy/core/src/multiarray/ctors.c
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: fixing

File: numpy/core/src/multiarray/ctors.c
Comment (mattip) [reply numpy/core/src/multiarray/ctors.c]: fixing

**Comment by mattip** [comment]
Squashed to one commit

**Comment by seberg** [comment]
Thanks matti, let me put this in. I feel the error message could be nicer, but do not have a good idea right now and its easy to fix up. (Will squash some whitespace ony changes, so not waiting for it).

**Comment by idanmiara** [comment]
Thanks for this fix! Just pointing out to whoever googled their way here later:
The `VisibleDeprecationWarning` was replaced by `ValueError` in numpy v1.24:
https://numpy.org/devdocs/release/1.24.0-notes.html

`ValueError: setting an array element with a sequence. The requested array has an inhomogeneous shape after 1 dimensions. The detected shape was (2,) + inhomogeneous part.`

**Pull request merged**
